### PR TITLE
feat(zarr): Flatten zarr module, remove XML-parsing

### PR DESCRIPTION
**Pull request by manzt**
Applies changes from #1446 to `master`

### CODE REVIEW

File: modules/zarr/src/index.ts [hunk 1]
```diff
@@ -1,4 +1,2 @@
 export { loadZarr } from './lib/load-zarr';
```
Comment (manzt) [inline modules/zarr/src/index.ts:1]: Single function

File: modules/zarr/src/lib/load-zarr.ts [hunk 2]
```diff
@@ -1,47 +1,24 @@
...
-  multiscales: Multiscale[];
-}
-
-export async function loadZarr(store: ZarrArray['store']) {
+export async function loadZarr(root: string | Store, options: ZarrOptions = {}) {
+  const store = normalizeStore(root);
   const { data, rootAttrs } = await loadMultiscales(store);
-  const labels = ['t', 'c', 'z', 'y', 'x'] as Labels<['t', 'c', 'z']>;
   const tileSize = guessTileSize(data[0]);
+  const labels = options.labels ?? guessLabels(rootAttrs);
```
Comment (manzt) [inline modules/zarr/src/lib/load-zarr.ts:13]: If no labels are provided, they are inferred (if possible) from the `rootAttrs`. For now, `guessLabels` only infers if OME-Zarr.

File: modules/zarr/src/lib/utils.ts [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    return new FetchFileStore(source);
+  }
+  return source;
+}
+
+export async function loadMultiscales(store: Store, path = '') {
+  const grp = await openGroup(store, path);
+  const rootAttrs = await grp.attrs.asObject() as RootAttrs;
+
+  if (!Array.isArray(rootAttrs.multiscales)) {
```
Comment (manzt) [inline modules/zarr/src/lib/utils.ts:19]: checks that multiscales extension is implemented in rootAttrs (generic across Zarr datasets)

File: modules/zarr/test/data/ome.zarr/.zattrs [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+                    "path": "0"
+                },
+                {
+                    "path": "1"
+                }
+            ],
+            "version": "0.1"
+        }
+    ],
+    "omero": {
```
Comment (manzt) [inline modules/zarr/test/data/ome.zarr/.zattrs:15]: OME metadata in root attrs. Used to distinguish multiscale dataset _is_ OME-Zarr.

**Comment by manzt** [comment]
Should be ready for review. I consolidated the handling of different Zarr formats in `loadZarr`, and added an OME-Zarr fixture.

**Comment by manzt** [comment]
The tests are failing in the browser:

```
fetch-file.js:17 GET http://localhost:5000/modules/zarr/test/data/multiscale.zarr/.zgroup 404 (Not Found)
```

The issue seems to be with `FetchFileStore` which wraps `fetchFile`.

**Comment by manzt** [comment]
Rebased again on master. I commented out the browser tests for zarr due to the issue with `fetchFile`/`FetchFileStore`. I'm not sure how to resolve. It seems that a zarr store wrapping `fetchFile` should work in browser and node environments.

**Review by kylebarron — APPROVED** [review]

File: modules/zarr/src/lib/storage.ts [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+import type {AsyncStore} from 'zarr/types/storage/types';
+
+import {joinUrlParts} from './utils';
+
+export class FetchFileStore implements AsyncStore<ArrayBuffer> {
+  constructor(public root: string) {}
+
+  async getItem(key: string, options?: RequestInit): Promise<ArrayBuffer> {
+    const filepath = joinUrlParts(this.root, key);
+    const response = await fetchFile(filepath, options);
```
[the remark this answers is not in the document]
Comment (manzt) [reply modules/zarr/src/lib/storage.ts:12]: The [`zarr.HTTPStore`](https://github.com/gzuidhof/zarr.js/blob/b55e4acf9b7b4f776036fe46a169ea91ad2ec79c/src/storage/httpStore.ts#L18) does exactly this ( uses `fetch` internally and expects a polyfil in node).

I guess the question is if `loadZarr` is given a string source, we need to pick the correct store to initialize `FileFetchStore` vs `HTTPStore` depending on the URL, correct?

File: modules/zarr/package.json [hunk 6]
```diff
@@ -15,7 +15,7 @@
     "loader",
     "zarr"
   ],
-  "types": "src/index.ts",
+  "types": "src/index.d.ts",
```
[the remark this answers is not in the document]
Comment (manzt) [reply modules/zarr/package.json:18]: Ah, right. Sorry. How is emitting types handled for other modules? Should I just point to the `src/index.ts`?

File: modules/zarr/src/lib/load-zarr.ts [hunk 7]
```diff
@@ -1,44 +1,24 @@
...
-}
+  if (!validLabels(labels, data[0].shape)) {
+    throw new Error('Invalid labels for Zarr array dimensions.');
+  }
 
-export async function loadZarr(store: ZarrArray['store']) {
-  const {data, rootAttrs} = await loadMultiscales(store);
-  const labels = ['t', 'c', 'z', 'y', 'x'] as Labels<['t', 'c', 'z']>;
-  const tileSize = guessTileSize(data[0]);
   const pyramid = data.map((arr) => new ZarrPixelSource(arr, labels, tileSize));
```
[the remark this answers is not in the document]
Comment (manzt) [reply modules/zarr/src/lib/load-zarr.ts:22]: This loads the `.zarray` metadata for all the arrays in the pyramid, but all ZarrPixelSource objects are completely lazy after the array metadata is loaded. There is no pixel/chunk data loaded.

File: modules/zarr/src/lib/utils.ts (hunk 3 above)
Comment (manzt) [reply modules/zarr/src/lib/utils.ts:21]: A Zarr store is organized into a hierarchy of nodes. Each node can either be a `group` or an `array`, and only a `group` may contain other nodes. The multiscales extension is _not_ OME-Zarr -specific (although it was spearheaded by OME folks working on zarr). 

The mutiscales extension adds a `"multiscales"` key-value to `group` metadata that ultimately provides the relative paths to other array sub-resolutions saved in the store. OME-Zarr chooses to also store additional metadata in the same group but under a different key (`omero`). This function reads the multiscales metadata (to initialize the pixel-sources), but simply returns the root metadata back to the end-user to choose what to do with it.

File: modules/zarr/src/lib/storage.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (manzt) [reply modules/zarr/src/lib/storage.ts:12]: Still running into issues here. How can do I resolve the alias for a fixture path to either a path or URL? e.g.

```javascript
const path = '@loaders.gl/zarr/test/data/multiscale.zarr';

// /Users/..../zarr/test/data/multiscale.zarr 
// or
// https://localhost:5001/zarr/test/data/multiscale.zarr
```

Am I missing something?

[the remark this answers is not in the document]
Comment (kylebarron) [reply modules/zarr/src/lib/storage.ts:12]: This alias should work

https://github.com/visgl/loaders.gl/blob/24e3e602099cc732a2f6590ee3435fe1e50d948c/test/aliases.js#L61

[the remark this answers is not in the document]
Comment (manzt) [reply modules/zarr/src/lib/storage.ts:12]: I'm missing something conceptually with how these aliases are resolved. For the following path:

```javascript
const path = '@loaders.gl/zarr/test/data/multiscale.zarr';
```

How can I resolve this to the HTTP endpoint the file is served from (`http://localhost:5000/modules/zarr/test/multiscale.zarr`) for use with the `fetch` API? Calling `resolvePath(path)` returns `/modules/zarr/test/multiscale.zarr`)

File: modules/zarr/package.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (kylebarron) [reply modules/zarr/package.json:18]: I think? Should be able to check one of the new TS modules

**Comment by manzt** [comment]
There is still an issue with `fetch`-ing zarr files in the browser that I can't seem get to the bottom of (See comment above). This seems to be specific to fetching assets in the test suite, but the bundled modules do run in the browser.

I have disabled the browser tests like geotiff in favor of merging this PR sooner rather than later to avoid having this PR go stale. We can follow up in subsequent PR to fix both issues.

File: modules/zarr/src/lib/utils.ts (hunk 3 above)
Comment (joshmoore) [reply modules/zarr/src/lib/utils.ts:20]: Sorry for the late addition, but there has been off and on discussion of moving the multiscales to a "super-interface" of OME-Zarr, i.e. a more generalized pyramidal image spec.

cf. https://github.com/pydata/xarray/issues/4118#issuecomment-873179375

**Pull request merged**